### DISCUSSION TITLE
JSBSim in Flight Gear

### DISCUSSION THREAD

**Discussion by matheusdarocha** [JSBSim-Team/jsbsim#658] [General]
Hello everyone,

I would like to ask if there is a way to insert a JSBSim maneuver in **Flight Gear**, for example:

  event name="Trim with elevator"
		condition>	simulation/sim-time-sec ge 0.71 </condition
		set name="ic/targetNlf" value="2.5"/
		set name="simulation/do_simple_trim" value="3"/
	notify/
   /event

The goal is to see the aircraft doing this maneuver

Is that possible?

Thanks in advance

**Comment by seanmcleod** [comment]
Any reason you've asked pretty much the same question twice in two separate issues, this issue and issue https://github.com/JSBSim-Team/jsbsim/discussions/657 ?

**Reply by matheusdarocha** [reply]
Hello Sean, Before I asked to do this in python and then I thought it could be in any software. I will erase the previous question

**Comment by seanmcleod** [comment]
Yes you can script maneuvers in a script file and in general JSBSim output can be sent to FlightGear via an `output` element, take a look at:

https://github.com/JSBSim-Team/jsbsim/blob/375f5be0fdfeb210f4cf66f938e3e6fd426c2966/data_output/flightgear.xml

**Reply by matheusdarocha** [reply]
thanks a lot Sean, but I still have some difficulty to implement this

The step by step of what I did was:
1. tried to open Flight gear by cmd using the command 
_fgfs.exe -fg-root="{your path}" -aircraft=c172p -native-fdm=socket,in,60,,5550,udp -fdm=external_
It didnt work and gave the answer that no -fg root or -aircraft is avaiable

then i put the cmd line in the Flight Gear correct path and opened it using
_fgfs.exe_

flight gear opened with C172P 

then i download the folder at the link, and tried the command at cmd 

src/jsbsim -realtime data_output/flightgear.xml scripts/c1723.xml

but it didnt work

I tried to go with cmd line to the folder where the script was, but it didnt work as well.

Do you have an example that i can use to try to do the maneuver?

Thanks a lot for the attention so far

Matt

**Comment by bcoconni** [comment]
There is [a detailed tutorial in our FAQ](https://github.com/JSBSim-Team/jsbsim/wiki/Frequently-Asked-Questions#how-can-i-interface-jsbsim-with-flightgear-).

Please follow the steps carefully, especially when asked to provide some options to the command line such as `--native-fdm=socket,in,60,,5550,udp`. If you skip any of them don't expect to get any working result.

If this example does not work correctly on your PC then please provide details about what is going wrong ; "it didnt work" is not very helpful.

**Reply by matheusdarocha** [reply]
thank you bcoconni, it work perfectly!

By the way, do you know any way to insert the maneuvers by python instead of a xml file?

**Comment by seanmcleod** [comment]
I posted some python based code examples in this disucssion - https://github.com/JSBSim-Team/jsbsim/discussions/519

**Reply by matheusdarocha** [reply]
Thanks Sean! I took a look at it, but i already have the python code, but i want to see the maneuver done in python at Flight gear, is there a way to do it?

**Comment by seanmcleod** [comment]
You have all the information at hand, just add this `output` element in the aircraft model's file which will output FlightGear packets as the model is run via python.

https://github.com/JSBSim-Team/jsbsim/blob/375f5be0fdfeb210f4cf66f938e3e6fd426c2966/data_output/flightgear.xml#L18-L19

**Reply by matheusdarocha** [reply]
Hello Sean, thanks for the answer
Unfortunatelly when I put this lines on my aircraft models, after run the python code, it doesn't move the flight gear model.

the script that i am trying to run is 

```python
fdm.do_trim(1)   

for t in np.arange(0,20,dt):
    fdm.run()
    if t < 0.01:
        fdm.do_trim(1)
```

trim the aircraft, and then loop through time.

When i run in prompt like the JSBSim:
_python Filename.py data_output/flightgear.xml --realtime --nice_

it does run JSBSim but doesn't make the simulation output goes to FG

**Comment by seanmcleod** [comment]
> Unfortunatelly when I put this lines on my aircraft models,

If you do something that doesn't work please show exactly what you did and/or the output.

For example, here is an example (albeit commented out) of an `output` element in the C172p aircraft model file. Show us the equivalent version that you added for FlightGear output. Use the triple back-ticks ``` mechanism to format it as xml and/or python when pasting snippets. 

https://github.com/JSBSim-Team/jsbsim/blob/33735db0fb1e4fdf47f253a4215dbe3bb63a8e92/aircraft/c172p/c172p.xml#L916-L941

Unless you `Filename.py` does something explicitly with the arguments `data_output/flightgear.xml --realtime --nice` nothing is going to be done to handle those arguments.

Which does bring up the point with regards to `--realtime`, so given your current Python example:

```python
for t in np.arange(0,20,dt):
    fdm.run()
```

This is going to run as fast as your PC can run it, i.e. much faster than real-time, so who knows maybe it transmitted all 20*dt frames to FlightGear in less than a second and you didn't even notice the changes in FlightGear.

**Reply by matheusdarocha** [reply]
Thanks again for the attention and patience. I 'll try to explain what I did.

First I followed bcoconni instruction at [a detailed tutorial in our FAQ](https://github.com/JSBSim-Team/jsbsim/wiki/Frequently-Asked-Questions#how-can-i-interface-jsbsim-with-flightgear-). Using my script and my model, it worked really well and I can see the maneuver at flight gear as soon as it finishes.

The script that I used was:

```xml
<?xml version="1.0" encoding="utf-8"?>
<?xml-stylesheet type="text/xsl" href="http://jsbsim.sf.net/JSBSimScript.xsl"?>
<runscript xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:noNamespaceSchemaLocation="http://jsbsim.sf.net/JSBSimScript.xsd"
    name="Cruise flight with steady turn in 737.">

  <description>
    This is a very simple script that trims the aircraft with running
    engines at altitude and runs out to 100 seconds. Some state data
    is printed out at ten second intervals. 
  </description>

 <use aircraft="mb339pan" initialize="reset01"/>

  <run start="0" end="15" dt="0.008333">

    <property value="0"> simulation/notify-time-trigger </property>

    <event name="Set engines running">
      <condition> simulation/sim-time-sec le 0.1 </condition>
      <set name="propulsion/engine[0]/set-running" value="1"/>
      <set name="propulsion/engine[1]/set-running" value="1"/>
      <notify/>
    </event>

  <event name="Initialization 4">
   <description>Set the engine controls</description>
   <condition>
     simulation/sim-time-sec GE 0.0
   </condition>
   <notify/>
   <set name="fcs/advance-cmd-norm[0]"   value="1.0"/>
   <set name="fcs/advance-cmd-norm[1]"   value="1.0"/>
   <set name="fcs/throttle-cmd-norm[0]"  value="0.85"/>
   <set name="fcs/throttle-cmd-norm[1]"  value="0.85"/>
   <set name="propulsion/set-running" value="-1"/>
  </event>

    <event name="Trim">
      <condition>
        simulation/sim-time-sec le 3
      </condition>
      <set name="simulation/do_simple_trim" value="1"/>
      <notify>
        <property>velocities/vc-kts</property>
        <property>velocities/vc-fps</property>
        <property>velocities/vt-fps</property>
        <property>attitude/phi-rad</property>
        <property>attitude/theta-rad</property>
        <property>attitude/psi-rad</property>
      </notify>
    </event>

	  <event name="Trim with elevator">
		<condition>	simulation/sim-time-sec ge 10 </condition>
		<set name="fcs/TrimWithHTorEle" value="1"/>
		<set name="ic/targetNlf" value=".5"/>
		<set name="simulation/do_simple_trim" value="3"/>
	<notify/>
   </event>
  </run>
</runscript>
```

Now I want to do the same thing, but not with a script using xml, but a python code that represents the maneuver.
So i did the code below:

```python
import jsbsim
import math
import numpy as np
import sys
sys.path.insert(1,'F:/OS/WORKING/200019_-_Turkish_Aerospace_Projeto_ASOJ/ENG/01_Aeronautics/09_Simulation/05_Tools/00_Python_library')
import akaer_aerolib_v7 as ak
import os


model_path = ('C:/tecnica/matheus.rocha')
modelpath = (model_path + '/200019_ASOJ_Green_FDM_V5_2')
model_name = '200019_ASOJ_aircraft'

Altitude = 6000
Veq= 400
Vtrue = math.sqrt(ak.Rho(0)/ak.Rho(Altitude))*Veq
Nlf = 1

MCA = '6J'
ak.WriteXml(modelpath + '//aircraft//', model_name, \
'MC' + MCA, 25, 'Green', False) ## comentar
fdm = jsbsim.FGFDMExec(modelpath)

fdm.load_model(model_name)
fdm['aero/blue'] = 1 ## comentar

dt = 0.001
fdm.set_dt(dt)

fdm['fcs/flap-pos-deg'] = 0
fdm['aero/alpha-max-rad'] = math.radians(20.0)
fdm['aero/alpha-min-rad'] = math.radians(-16.0)
fdm['ic/h-agl-ft'] = Altitude
fdm['ic/h-sl-ft'] = Altitude
fdm['ic/vt-kts'] = Vtrue

fdm.do_trim(1)   

for t in np.arange(0,20,dt):
    fdm.run()
    if t < 0.01:
        fdm.do_trim(1)   
        
print('end')
```
   
The aircraft model is the same, but I added the output line as you said.
It runs really well, just as the xml file, but the connection between python and FG does not work, it does not bring any error, just does not move to trim Nz = 1

**Reply by bcoconni** [reply]
@matheusdarocha, you need to specify the log directive `data_output/flightgear.xml` to your FDM. So your script should look like:
```python
fdm = jsbsim.FGFDMExec(modelpath)

fdm.load_model(model_name)
fdm.set_output_directive('path/to/jsbsim_root/data_output/flightgear.xml')  # Tell JSBSim to send data to FlightGear.

fdm['aero/blue'] = 1 ## comentar
```
Note the additional code line:
```python
fdm.set_output_directive('path/to/jsbsim_root/data_output/flightgear.xml')
```
Of course you should replace the path `path/to/jsbsim_root` by the appropriate path so that JSBSim can load the file `data_output/flightgear.xml`.

**Reply by matheusdarocha** [reply]
@bcoconni thanks for your attention

I tried to put the additional line as you said but the aircraft at FG didn't move.
What I did was:

start FG using command line at prompt:

```sh
fgfs --fdm=null --native-fdm=socket,in,60,,5550,udp --fdm=external --aircraft=mb339pan
```

run the code in python below

```python
import jsbsim
import math
import numpy as np
import sys
sys.path.insert(1,'F:/OS/WORKING/200019_-_Turkish_Aerospace_Projeto_ASOJ/ENG/01_Aeronautics/09_Simulation/05_Tools/00_Python_library')
import akaer_aerolib_v7 as ak
import os

model_path = ('C:/tecnica/matheus.rocha')
modelpath = (model_path + '/200019_ASOJ_Green_FDM_V5_2')
model_name = '200019_ASOJ_aircraft'

Altitude = 6000
Veq= 400
Vtrue = math.sqrt(ak.Rho(0)/ak.Rho(Altitude))*Veq
Nlf = 1

MCA = '6J'
ak.WriteXml(modelpath + '//aircraft//', model_name, \
'MC' + MCA, 25, 'Green', False) ## comentar
fdm = jsbsim.FGFDMExec(modelpath)

fdm.load_model(model_name)

fdm.set_output_directive('data_output/flightgear.xml')  # Tell JSBSim to send data to FlightGear.

fdm['aero/blue'] = 1 ## comentar

dt = 0.001
fdm.set_dt(dt)

fdm['fcs/flap-pos-deg'] = 0

fdm['aero/alpha-max-rad'] = math.radians(20.0)
fdm['aero/alpha-min-rad'] = math.radians(-16.0)

fdm['ic/h-agl-ft'] = Altitude
fdm['ic/h-sl-ft'] = Altitude
fdm['ic/vt-kts'] = Vtrue

fdm.do_trim(1)   

for t in np.arange(0,20,dt):
    fdm.run()
    if t < 0.01:
        fdm.do_trim(1)   
        
print('end')
```

and then nothing happens on flight gear

**Comment by seanmcleod** [comment]
> but I added the output line as you said.

Can you add a snippet of the xml to show exactly where and what you added to the aircraft's xml file?

What happens if you add an `output` element like the one I showed from the C-172p example, i.e. one that outputs a CSV file when you run your python script? Does the CSV file get created and populated?

**Reply by matheusdarocha** [reply]
@seanmcleod thanks for the attention

The xml that i put the addional line is below 

```xml
<?xml version="1.0"?>
    <?xml-stylesheet type="text/xsl" href="http://jsbsim.sourceforge.net/JSBSim.xsl"?>
    <fdm_config name="m_ASOJ_global6000_green_v01" release="BETA" version="2.0"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="http://jsbsim.sourceforge.net/JSBSim.xsd">
        
        <fileheader>
            <author> Akaer </author>
            <filecreationdate> 2022-02-02 </filecreationdate>
            <version>$Revision: 1.00 $</version>
            <description> Bombardier Global 6000 - ASOJ Green Aircraft </description>
            <license>
                <licenseName>GPL (General Public License)</licenseName>
                <licenseURL>http://www.gnu.org/licenses/gpl.html</licenseURL>
            </license>
            <note>
                This model may contain confidential and proprietary material for the 
                sole use of the intended recipient and/or may be subject to export 
                control regulations. Any review or distribution by others is strictly 
                prohibited. If you are not the intended recipient, please contact 
                Akaer at +55 (12) 2139-1100 or https://www.akaer.com.br/eng/contact
          </note>
        </fileheader>
    	
    	<metrics file="metrics/200019_ASOJ_Green_Coeff_metrics"/>
    <!-- -->
    	<mass_balance file="mass_balance/200019_ASOJ_Green_CG25_MC6J"/>
    
    <!-- -->
    <!-- Include engine data and fuel mass configuration -->
    	<propulsion file="propulsion/200019_ASOJ_Green_Engine_BR710_CG25_MC6J"/>
    	
    		<!-- Include ground data -->
        <ground_reactions>
    		<!-- Include ground parameters-->
        </ground_reactions>
    	<system name="BlueAircraftActive">
    		<channel name="BlueAircraft">	
    			<switch name="aero/blue">
    				<default value="0.0"/>
    				<test value="aero/blue">
    					aero/blue eq 1.0
    				</test>	
    			</switch>
    		</channel>
    	</system>
    	
    	<system file="metrics/200019_ASOJ_Coeff_GeoParam"/>
    
    	<system file="200019_ASOJ_FCS_Flap"/>
    	
    	<system file="200019_ASOJ_FCS_Pitch"/>
    	
    	<system file="200019_ASOJ_FCS_Roll_v3_2"/>
    	
    	<system file="200019_ASOJ_FCS_Yaw"/>
    	
    	<system file="metrics/200019_ASOJ_Coeff_Arms_v04"/>
    	
    
    	<aerodynamics file="aerodynamics/200019_ASOJ_AerodynamicDB"/>
    	
        <output name="localhost" type="FLIGHTGEAR" protocol="UDP" port="5550" rate="60" version="24"> 
 </output> 
    </fdm_config>
    <!-- -->
```

start FG using command line at prompt:

```sh
fgfs --fdm=null --native-fdm=socket,in,60,,5550,udp --fdm=external --aircraft=mb339pan
```

run the code in python below

```python
import jsbsim
import math
import numpy as np
import sys
sys.path.insert(1,'F:/OS/WORKING/200019_-_Turkish_Aerospace_Projeto_ASOJ/ENG/01_Aeronautics/09_Simulation/05_Tools/00_Python_library')
import akaer_aerolib_v7 as ak
import os

model_path = ('C:/tecnica/matheus.rocha')
modelpath = (model_path + '/200019_ASOJ_Green_FDM_V5_2')
model_name = '200019_ASOJ_aircraft'

Altitude = 6000
Veq= 400
Vtrue = math.sqrt(ak.Rho(0)/ak.Rho(Altitude))*Veq
Nlf = 1

MCA = '6J'
ak.WriteXml(modelpath + '//aircraft//', model_name,
'MC' + MCA, 25, 'Green', False) ## comentar
fdm = jsbsim.FGFDMExec(modelpath)

fdm.load_model(model_name)

fdm.set_output_directive('data_output/flightgear.xml') # Tell JSBSim to send data to FlightGear.

fdm['aero/blue'] = 1 ## comentar

dt = 0.001
fdm.set_dt(dt)

fdm['fcs/flap-pos-deg'] = 0

fdm['aero/alpha-max-rad'] = math.radians(20.0)
fdm['aero/alpha-min-rad'] = math.radians(-16.0)

fdm['ic/h-agl-ft'] = Altitude
fdm['ic/h-sl-ft'] = Altitude
fdm['ic/vt-kts'] = Vtrue

fdm.do_trim(1)

for t in np.arange(0,20,dt):
fdm.run()
if t < 0.01:
fdm.do_trim(1)

print('end')
```

and then nothing happens on flight gear

About the CSV file with the outputs, it is usually not created when run with python script, so it was not created

**Reply by seanmcleod** [reply]
Please use the triple back tick options, followed by the content type, e.g. `python`, `xml`, `sh` to format the relevant sections of your comment, makes it a lot easier to read.

> About the CSV file with the outputs, it is usually not created when run with python script, so it was not created

The point was to confirm whether `output` elements in your FDM were working at all, i.e. replace the Flight Gear specific `output` element with a simple CSV one and confirm whether that works, i.e. when running your python script the CSV file is created with data logged etc.

**Reply by matheusdarocha** [reply]
I changed the output to get a csv and it worked.
Also, I discovered, I had to reset to initial conditions before the loop, now I can run python and see the maneuver at JSBSim

Thanks a lot for all the help

**Reply by seanmcleod** [reply]
> now I can run python and see the maneuver at JSBSim

Do you mean you can see it in FlightGear now?

**Reply by matheusdarocha** [reply]
Yes I Can, it worked. Now i am working at @bcoconni comment to make it run in realtime
Thanks a lot

**Comment by bcoconni** [comment]
@matheusdarocha If you drive FlightGear from a Python script, you'll most likely notice that your aircraft flies much too fast. This is because JSBSim runs its computation as fast as possible i.e. much faster than real time.

If you want your simulation to run in real time, you'll need to wait that enough real time has elapsed since the last simulated time step. One means to achieve this result is the code below (the real time management is below the `if args.realtime:` instruction):
https://github.com/JSBSim-Team/jsbsim/blob/a119eb2e81b935f562e84480bb21c53ceb541eb7/python/JSBSim.py#L126-L139
